lib/elixir/src/elixir_bitstring.erl

```diff
@@ -244,12 +244,19 @@ validate_spec(signed, [])    -> {sign, signed};
 validate_spec(unsigned, [])  -> {sign, unsigned};
 validate_spec(_, _)          -> none.
 
-expand_spec_arg(Expr, S, _OriginalS, E) when is_atom(Expr); is_integer(Expr) ->
+expand_spec_arg(_, Expr, S, _OriginalS, E) when is_atom(Expr); is_integer(Expr) ->
   {Expr, S, E};
-expand_spec_arg(Expr, S, _OriginalS, #{context := match} = E) ->
-  {EExpr, SE, EE} = elixir_expand:expand(Expr, S#elixir_ex{prematch=raise}, E#{context := nil}),
+expand_spec_arg(size, Expr, S, _OriginalS, #{context := match} = E) ->
+  {EExpr, SE, EE} = elixir_expand:expand(Expr, S#elixir_ex{prematch=raise, bitsize=true}, E#{context := guard}),
   {EExpr, SE#elixir_ex{prematch=S#elixir_ex.prematch}, EE#{context := match}};
-expand_spec_arg(Expr, S, OriginalS, E) ->
+expand_spec_arg(size, Expr, S, OriginalS, E) ->
+  NewS = elixir_env:reset_read(S, OriginalS),
+  {EExpr, SE, EE} = elixir_expand:expand(Expr, NewS#elixir_ex{bitsize=true}, E#{context := guard}),
+  {EExpr, SE, EE#{context := nil}};
+expand_spec_arg(unit, Expr, S, _OriginalS, #{context := match} = E) ->
```

Comment:
Decided to keep the original clauses for the `unit` modifier case... so changes in this PR only affect `size` modifier (which was my original intention)